File: dotnet/src/Rerout/Models/CreateLinkInput.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Rerout.Models;

/// <summary>Request body for the <c>POST /v1/links</c> endpoint.</summary>
public sealed record CreateLinkInput
{
    /// <summary>Absolute <c>https://</c> destination URL. Max 2048 characters.</summary>
    [JsonPropertyName("target_url")]
    public required string TargetUrl { get; init; }

    /// <summary>Verified custom domain to host this link on. Omit for <c>rerout.co/:code</c>.</summary>
    [JsonPropertyName("domain_hostname")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? DomainHostname { get; init; }

    /// <summary>Custom path. Only valid when <see cref="DomainHostname"/> is supplied.</summary>
    [JsonPropertyName("code")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Code { get; init; }

    /// <summary>Unix seconds — expiration. Omit for a permanent link.</summary>
    [JsonPropertyName("expires_at")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? ExpiresAt { get; init; }

    /// <summary>Override social preview title. Max 90 characters.</summary>
    [JsonPropertyName("seo_title")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? SeoTitle { get; init; }

    /// <summary>Override social preview description. Max 220 characters.</summary>
    [JsonPropertyName("seo_description")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? SeoDescription { get; init; }

    /// <summary>Absolute <c>https://</c> social preview image URL.</summary>
    [JsonPropertyName("seo_image_url")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? SeoImageUrl { get; init; }

    /// <summary>Canonical URL for the preview HTML.</summary>
    [JsonPropertyName("seo_canonical_url")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? SeoCanonicalUrl { get; init; }

    /// <summary>Whether the preview page should be marked <c>noindex</c>.</summary>
    [JsonPropertyName("seo_noindex")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? SeoNoindex { get; init; }

    /// <summary>Smart Links — plaintext password to gate the link. Hashed server-side.</summary>
    [JsonPropertyName("password")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Password { get; init; }

    /// <summary>Smart Links — cap the link to this many clicks.</summary>
    [JsonPropertyName("max_clicks")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? MaxClicks { get; init; }

    /// <summary>Smart Links — mint a conversion click id on redirect.</summary>
    [JsonPropertyName("track_conversions")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? TrackConversions { get; init; }

    /// <summary>Smart Links — ordered geo/device routing rules (full set).</summary>
    [JsonPropertyName("routing_rules")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public IReadOnlyList<RoutingRule>? RoutingRules { get; init; }

    /// <summary>Smart Links — weighted A/B destinations (full set).</summary>
    [JsonPropertyName("ab_variants")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public IReadOnlyList<AbVariantInput>? AbVariants { get; init; }
}

File: ruby/lib/rerout/create_link_input.rb
# frozen_string_literal: true

module Rerout
  # Request body for `POST /v1/links`. Only `target_url` is required —
  # everything else is optional and omitted from the payload when not set.
  class CreateLinkInput
    attr_reader :target_url, :domain_hostname, :code, :expires_at,
                :seo_title, :seo_description, :seo_image_url,
                :seo_canonical_url, :seo_noindex,
                :password, :max_clicks, :track_conversions,
                :routing_rules, :ab_variants

    # @param target_url [String] required, the destination URL.
    # @param domain_hostname [String, nil] verified custom domain (e.g. `go.brand.com`).
    # @param code [String, nil] custom path. Only allowed with a verified `domain_hostname`.
    # @param expires_at [Integer, nil] unix seconds.
    # @param seo_title [String, nil]
    # @param seo_description [String, nil]
    # @param seo_image_url [String, nil] absolute https:// URL.
    # @param seo_canonical_url [String, nil]
    # @param seo_noindex [Boolean, nil] default server-side: `true`.
    # @param password [String, nil] gate the link behind a password.
    # @param max_clicks [Integer, nil] disable the link after this many clicks.
    # @param track_conversions [Boolean, nil] enable conversion tracking.
    # @param routing_rules [Array<Rerout::Models::RoutingRule, Hash>, nil]
    #   Smart Link routing rules.
    # @param ab_variants [Array<Rerout::Models::AbVariant, Hash>, nil]
    #   A/B-test variants — each `{ target_url:, weight? }`.
    def initialize(target_url:, domain_hostname: nil, code: nil, expires_at: nil,
                   seo_title: nil, seo_description: nil, seo_image_url: nil,
                   seo_canonical_url: nil, seo_noindex: nil,
                   password: nil, max_clicks: nil, track_conversions: nil,
                   routing_rules: nil, ab_variants: nil)
      raise ArgumentError, 'target_url is required' if target_url.nil? || target_url.to_s.empty?

      @target_url = target_url
      @domain_hostname = domain_hostname
      @code = code
      @expires_at = expires_at
      @seo_title = seo_title
      @seo_description = seo_description
      @seo_image_url = seo_image_url
      @seo_canonical_url = seo_canonical_url
      @seo_noindex = seo_noindex
      @password = password
      @max_clicks = max_clicks
      @track_conversions = track_conversions
      @routing_rules = routing_rules
      @ab_variants = ab_variants
      freeze
    end

    # Serialize for the wire. Fields are only included when set.
    def to_h
      hash = { 'target_url' => target_url }
      hash['domain_hostname'] = domain_hostname unless domain_hostname.nil?
      hash['code'] = code unless code.nil?
      hash['expires_at'] = expires_at unless expires_at.nil?
      hash['seo_title'] = seo_title unless seo_title.nil?
      hash['seo_description'] = seo_description unless seo_description.nil?
      hash['seo_image_url'] = seo_image_url unless seo_image_url.nil?
      hash['seo_canonical_url'] = seo_canonical_url unless seo_canonical_url.nil?
      hash['seo_noindex'] = seo_noindex unless seo_noindex.nil?
      hash['password'] = password unless password.nil?
      hash['max_clicks'] = max_clicks unless max_clicks.nil?
      hash['track_conversions'] = track_conversions unless track_conversions.nil?
      hash['routing_rules'] = routing_rules.map { |r| InputSerialization.rule_hash(r) } unless routing_rules.nil?
      hash['ab_variants'] = ab_variants.map { |v| InputSerialization.variant_hash(v) } unless ab_variants.nil?
      hash
    end
  end
end
